utils/fork-tree/src/lib.rs

```diff
@@ -377,36 +382,57 @@ where
 				None => return Ok(RemovedIterator { stack: Vec::new() }),
 			};
 
-		let mut old_roots = std::mem::take(&mut self.roots);
+		let mut removed = std::mem::take(&mut self.roots);
 
-		let curr_children = root_index
+		// Find and detach the new root from the removed nodes
+		let root_siblings = root_index
 			.iter()
 			.take(root_index.len() - 1)
-			.fold(&mut old_roots, |curr, idx| &mut curr[*idx].children);
-		let mut root = curr_children.remove(root_index[root_index.len() - 1]);
-
-		let mut removed = old_roots;
+			.fold(&mut removed, |curr, idx| &mut curr[*idx].children);
+		let root = root_siblings.remove(root_index[root_index.len() - 1]);
+		self.roots = vec![root];
 
-		// we found the deepest ancestor of the finalized block, so we prune
-		// out any children that don't include the finalized block.
-		let root_children = std::mem::take(&mut root.children);
-		let mut is_first = true;
+		// If, because of the `predicate`, the new root is not the deepest
+		// ancestor of `hash` then we can remove all the nodes that are
+		// descendants of the new `root` but not ancestors of `hash`.
+		let mut curr = &mut self.roots[0];
+		loop {
+			let mut next_ancestor_idx = usize::MAX;
+			for (idx, child) in curr.children.iter().enumerate() {
+				if child.number == *number && child.hash == *hash ||
+					child.number < *number && is_descendent_of(&child.hash, hash)?
+				{
+					next_ancestor_idx = idx;
+					break
+				}
+			}
+			if next_ancestor_idx == usize::MAX {
+				// We are above all current node children, stop searching
+				break
+			}
+			// Preserve only the ancestor node, the siblings are removed
+			let mut next_siblings = std::mem::take(&mut curr.children);
+			let next = next_siblings.remove(next_ancestor_idx);
+			curr.children = vec![next];
+			removed.append(&mut next_siblings);
+			curr = &mut curr.children[0];
+		}
 
-		for child in root_children {
-			if is_first &&
-				(child.number == *number && child.hash == *hash ||
-					child.number < *number && is_descendent_of(&child.hash, hash)?)
-			{
-				root.children.push(child);
-				// assuming that the tree is well formed only one child should pass this
-				// requirement due to ancestry restrictions (i.e. they must be different forks).
-				is_first = false;
-			} else {
-				removed.push(child);
+		// Curr now points to our direct ancestor, if necessary remove any node that is
+		// not a descendant of `hash`.
+		if curr.children.len() > 0 {
+			let children = std::mem::take(&mut curr.children);
+			for child in children {
+				if child.number == *number && child.hash == *hash ||
+					*number < child.number && is_descendent_of(hash, &child.hash)?
```

Comment:
nit: Maybe one comment here that this condition is inverted and we are now searching for the prune "target" or a node even deeper than the "target".